### PR TITLE
indents(hcl): no indentation after brace

### LINKED ISSUE #2590: Auto indent outside blocks for hcl
### Describe the bug

When adding new line auto indent works great when inside a block, but when outside of a block it also auto indent when new line is added.

**Screenshots**
Hard to show on screenshots. But here is an attempt. See it on line 5.
image

### To Reproduce

1. I use stock [AstroVim](https://github.com/kabinspace/AstroVim) with :TSInstall hcl
2. Write any Terraform block and try inserting a new line after that block. And see it auto indent.

### Expected behavior

I expect auto indent not to happen out side Terraform blocks.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.4 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v17.6.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (GCC) 11.2.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - hcl            ✓ . ✓ ✓ ✓ 
  - bash           ✓ ✓ ✓ . ✓ 
  - yaml           ✓ ✓ ✓ ✓ ✓ 
  - lua            ✓ ✓ ✓ ✓ ✓ 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.6.1
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by builduser

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Additional context

 Also disabling indent makes it work outside Terraform blocks, but breaks indentation inside blocks. This issue is moved up to this project by suggestion of the maintainer of this project https://github.com/MichaHoffmann/tree-sitter-hcl/issues/15.

### PR DESCRIPTION
**Pull request by theHamsta**
Fixes #2590

### CODE REVIEW

**Comment by theHamsta** [comment]
No idea why tests would also pass without the fix. Is the parser really used in the tests?

**Comment by theHamsta** [comment]
@MichaHoffmann I think more rules are required to make indents in HCL work as expected. Can you please check whether more adaptions are required? You can get inspirations from the indent file for C

File: tests/indent/hcl_spec.lua [hunk 1]
```diff
@@ -0,0 +1,20 @@
+local Runner = require("tests.indent.common").Runner
+--local XFAIL = require("tests.indent.common").XFAIL
+
+local run = Runner:new(it, "tests/indent/hcl", {
+  tabstop = 2,
+  shiftwidth = 2,
+  expandtab = true,
+})
+
+describe("indent Lua:", function()
```
Comment (MichaHoffmann) [inline tests/indent/hcl_spec.lua:10]: should be HCL

Comment (theHamsta) [reply tests/indent/hcl_spec.lua:10]: :facepalm:

Comment (MichaHoffmann) [reply tests/indent/hcl_spec.lua]: The test failure can be found when searching for "indent Lua:" so the test works as expected!

**Comment by MichaHoffmann** [comment]
The tree generated for 
```
locals {
  titles = ["test0", "test1"]
}
```
is 
```
(config_file [0, 0] - [3, 0]
  (body [0, 0] - [2, 1]
    (block [0, 0] - [2, 1]
      (identifier [0, 0] - [0, 6])
      (block_start [0, 7] - [0, 8])
      (body [1, 2] - [1, 29]
        (attribute [1, 2] - [1, 29]
          (identifier [1, 2] - [1, 8])
          (expression [1, 11] - [1, 29]
            (collection_value [1, 11] - [1, 29]
              (tuple [1, 11] - [1, 29]
                (tuple_start [1, 11] - [1, 12])
                (expression [1, 12] - [1, 19]
                  (literal_value [1, 12] - [1, 19]
                    (string_lit [1, 12] - [1, 19]
                      (quoted_template_start [1, 12] - [1, 13])
                      (template_literal [1, 13] - [1, 18])
                      (quoted_template_end [1, 18] - [1, 19]))))
                (expression [1, 21] - [1, 28]
                  (literal_value [1, 21] - [1, 28]
                    (string_lit [1, 21] - [1, 28]
                      (quoted_template_start [1, 21] - [1, 22])
                      (template_literal [1, 22] - [1, 27])
                      (quoted_template_end [1, 27] - [1, 28]))))
                (tuple_end [1, 28] - [1, 29]))))))
      (block_end [2, 0] - [2, 1]))))

```

should we maybe just `indent_end`  at `block_end`?

File: tests/indent/hcl_spec.lua [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+describe("indent HCL:", function()
+  describe("whole file:", function()
+    run:whole_file(".", {
+      expected_failures = {},
+    })
+  end)
+
+  describe("new line:", function()
+    run:new_line("no-indent-after-brace.tf", { on_line = 4, text = "# Wow, no indent here please", indent = 2 })
```
Comment (MichaHoffmann) [inline tests/indent/hcl_spec.lua:18]: indent should be 0 here right?

Comment (theHamsta) [reply tests/indent/hcl_spec.lua:18]: yes, the 2 was just to try to make it failing.

Comment (MichaHoffmann) [reply tests/indent/hcl_spec.lua]: ah ok, fair enough, sorry

Comment (theHamsta) [reply tests/indent/hcl_spec.lua]: No, you were right. I forgot to remove that. Thank you!

File: queries/hcl/indents.scm [hunk 3]
```diff
@@ -11,3 +11,5 @@
   (block_end)
   (tuple_end)
 ] @branch
+
+(block_end) @indent_end
```
Comment (kyazdani42) [inline queries/hcl/indents.scm:15]: this does not work, i've tried it yesterday and when typing `o` it still indents at 2 after the block ends

Comment (MichaHoffmann) [reply queries/hcl/indents.scm:15]: the query matches @branch and @indent_end ; im really not sure how indentation works in nvim-treesitter tbh but maybe thats a problem? Should we remove block_end from @branch?

Comment (MichaHoffmann) [reply queries/hcl/indents.scm:15]: also we probably should start indenting when we see `block_start`

Comment (theHamsta) [reply queries/hcl/indents.scm:15]: It works for me.

Comment (MichaHoffmann) [reply queries/hcl/indents.scm:15]: > It works for me.

Oh thats odd, do you know why it fails in the CI run? It worked on an older commit if i remember correctly.

**Comment by MichaHoffmann** [comment]
Hey, is there something i can do to help out with this PR ?

**Comment by kyazdani42** [comment]
@MichaHoffmann maybe you can try to make the test pass :)

**Comment by MichaHoffmann** [comment]
> @MichaHoffmann maybe you can try to make the test pass :)

If @theHamsta is ok with it, i can try !

**Comment by MichaHoffmann** [comment]
using the queries 
```
[
  (object_start)
  (block_start)
  (tuple_start)
] @indent

[
  (object_end)
  (block_end)
  (tuple_end)
] @dedent
```

it kinda works: 
```
========================================	
Testing: 	/var/home/mhoffm/Git/Personal/nvim-treesitter/tests/indent/hcl_spec.lua	
Success	||	indent HCL: whole file: ./indent-in-multiline-tuples.tf	
Success	||	indent HCL: whole file: ./no-indent-after-brace.tf	
Success	||	indent HCL: new line: no-indent-after-brace.tf[4]	
Success	||	indent HCL: new line: indent-in-multiline-tuples.tf[4]	
	
Success: 	4	
Failed : 	0	
Errors : 	0	
```
i also added another test, to be a little more sure ( and get used to the indentation tests ):
```
diff --git a/tests/indent/hcl/indent-in-multiline-tuples.tf b/tests/indent/hcl/indent-in-multiline-tuples.tf
new file mode 100644
index 0000000..4024878
--- /dev/null
+++ b/tests/indent/hcl/indent-in-multiline-tuples.tf
@@ -0,0 +1,6 @@
+test {
+  x = [
+    1,
+    2,
+  ]
+}
diff --git a/tests/indent/hcl/no-indent-after-brace.tf b/tests/indent/hcl/no-indent-after-brace.tf
new file mode 100644
index 0000000..e670ad8
--- /dev/null
+++ b/tests/indent/hcl/no-indent-after-brace.tf
@@ -0,0 +1,4 @@
+# Issue #2590
+locals {
+  titles = ["test0", "test1"]
+}
diff --git a/tests/indent/hcl_spec.lua b/tests/indent/hcl_spec.lua
new file mode 100644
index 0000000..a62547a
--- /dev/null
+++ b/tests/indent/hcl_spec.lua
@@ -0,0 +1,23 @@
+local Runner = require("tests.indent.common").Runner
+--local XFAIL = require("tests.indent.common").XFAIL
+
+local run = Runner:new(it, "tests/indent/hcl", {
+  tabstop = 2,
+  shiftwidth = 2,
+  expandtab = true,
+})
+
+describe("indent HCL:", function()
+  describe("whole file:", function()
+    run:whole_file(".", {
+      expected_failures = {},
+    })
+  end)
+
+  describe("new line:", function()
+    run:new_line("no-indent-after-brace.tf", { on_line = 4, text = "# Wow, no indent here please", indent = 0 })
+  end)
+  describe("new line:", function()
+    run:new_line("indent-in-multiline-tuples.tf", { on_line = 4, text = "3,", indent = 4 })
+  end)
+end)
```

**Comment by MichaHoffmann** [comment]
I had to add a `ftdetect` to make neovim use the parser in my local setup though.

**Comment by kyazdani42** [comment]
@MichaHoffmann which version of neovim are you using ? It works fine on 0.8. Ftdetects where removed from nvim-treesitter not so long ago because they don't make sense here :)

**Comment by MichaHoffmann** [comment]
> @MichaHoffmann which version of neovim are you using ? It works fine on 0.8. Ftdetects where removed from nvim-treesitter not so long ago because they don't make sense here :)

Oh i guess im on an ancient version and should update sometime. I'm on `NVIM v0.6.1

**Comment by kyazdani42** [comment]
we are going to drop support for 0.6 when 0.7 gets released at the end of the month :)

**Comment by MichaHoffmann** [comment]
> we are going to drop support for 0.6 when 0.7 gets released at the end of the month :)

Im updating as we speak! Anyway, the queries work better like i described. Once this is merged ill write a few more tests

**Comment by kyazdani42** [comment]
can you create a separate PR ? we'll close this one then

**Comment by MichaHoffmann** [comment]
> can you create a separate PR ? we'll close this one then

Ok, i'll do that sometime today!

**Comment by theHamsta** [comment]
Superseded by #2855

**Pull request closed without merging**